wheatley/row_generation/complib_composition_generator.py

```diff
@@ -110,34 +158,9 @@ def process_call_string(calls: str) -> List[str]:
         super().__init__(response_rows['stage'])
 
     @classmethod
-    def from_url(cls, url: str) -> RowGenerator:
-        """ Generates a ComplibCompositionGenerator from a URL to that composition. """
-        # ==== CANONICALISATION ====
-        # Always strip the 'https://' and 'www.' from the front of the URL
-        canonical_url = url
-        canonical_url = removeprefix(canonical_url, "https://")
-        canonical_url = removeprefix(canonical_url, "www.")
-
-        parts = canonical_url.split("?accessKey=")
-        main_url = parts[0]
-        try:
-            access_key: Optional[str] = parts[1]
-        except IndexError:
-            access_key = None
-
-        main_url_parts = main_url.split("/")
-
-        try:
-            if not main_url_parts[0].endswith("complib.org"):
-                raise InvalidComplibURLError(url, "Doesn't point to 'complib.org'.")
-            if main_url_parts[1] != "composition":
-                raise InvalidComplibURLError(url, "Not a composition.")
-            comp_id = int(main_url_parts[2])
-        except KeyError as e:
-            raise InvalidComplibURLError(url, "URL has too few segments.") from e
-        except ValueError as e:
-            raise InvalidComplibURLError(url, f"ID {main_url_parts[2]} is not an integer.") from e
-
+    def from_arg(cls, arg: str) -> RowGenerator:
```

Comment:
```suggestion
    def from_arg(cls, arg: str) -> 'ComplibCompositionGenerator':
```
would make type checking in unit tests happier

Reply:
Wouldn't this break the type checking in the `wheatley` module itself?  Because the return value will have type specified by the `cls` argument, which can't be relied upon to be `ComplibCompositionGenerator`.  At any rate, I don't think it's worth running any kind of static analysis on the unit tests - they don't really need it and anyway have to do things like access private fields of other classes.  So I'd rather have correct type checking of the main library, as opposed to making the tests' static analysis happy.

Reply:
Also, couldn't we annotate the unit tests with the specific type?  I think that's allowed...

Reply:
A string is just because you can't refer to the class that's being defined. `Bell` does the same thing for its factory methods. https://www.python.org/dev/peps/pep-0484/#forward-references

The type `cls` being used here will be a type derived from `ComplibCompositionGenerator` (unless you do something really odd), so it's just a more accurate annotation.

It just is odd to open it in an IDE and have warnings / errors in the unit tests which can be avoided.

Reply:
Oh!  OK I get what's going on here - it makes sense now.